src/language_features/document_links.rs

```diff
@@ -0,0 +1,213 @@
+                .map(|v| (server_id, v))
+                .collect();
+            v
+        })
+        .collect::<Vec<_>>();
+    links.sort_by_key(|(_, link)| link.range.start);
+
+    let buffile = &meta.buffile;
+    let document = match ctx.documents.get(buffile) {
+        Some(document) => document,
+        None => {
+            ctx.document_links.remove(buffile);
+            return;
+        }
+    };
+    let version = document.version;
+    let range_specs = links
+        .iter()
+        .map(|(server_id, link)| {
+            let server = ctx.server(*server_id);
+            format!(
+                "{}|LspDocumentLink",
+                lsp_range_to_kakoune(&link.range, &document.text, server.offset_encoding)
+            )
+        })
+        .join(" ");
+
+    ctx.document_links.insert(buffile.clone(), links);
+
+    let command = format!("set-option buffer lsp_document_links {version} {range_specs}");
+    let command = format!(
+        "evaluate-commands -buffer {} %§{}§",
+        editor_quote(buffile),
+        command.replace('§', "§§")
+    );
+    ctx.exec(EditorMeta::default(), command);
+}
+
+#[derive(Clone, Debug)]
+pub struct DocumentLinkOptions {
+    pub selection_desc: String,
+}
+
+pub fn resolve_and_follow_document_link(
+    meta: EditorMeta,
+    params: DocumentLinkOptions,
+    ctx: &mut Context,
+) {
+    let (range, _cursor) = parse_kakoune_range(&params.selection_desc);
+    let document = match ctx.documents.get(&meta.buffile) {
+        Some(document) => document,
+        None => return,
+    };
+
+    if let Some((server_id, link)) = ctx
+        .document_links
+        .get(&meta.buffile)
+        .and_then(|links| {
+            links.iter().find(|(server_id, link)| {
+                let ServerSettings {
+                    offset_encoding, ..
+                } = ctx.server(*server_id);
+                let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+                ranges_overlap(link.range, range)
+            })
+        })
+        .filter(|(_, link)| link.target.is_none())
+        .cloned()
+    {
+        let mut req_params = HashMap::new();
+        req_params.insert(server_id, vec![link]);
+
+        ctx.call::<DocumentLinkResolve, _>(
+            meta,
+            RequestParams::Each(req_params),
+            |ctx: &mut Context, meta, results| follow_document_link(meta, &results, ctx),
+        );
+        return;
+    }
+
+    let no_links = vec![];
+    let links = ctx.document_links.get(&meta.buffile).unwrap_or(&no_links);
+    let mut links = links
+        .iter()
+        .filter(|(server_id, link)| {
+            let ServerSettings {
+                offset_encoding, ..
+            } = ctx.server(*server_id);
+            let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+            ranges_overlap(link.range, range)
+        })
+        .map(|(a, b)| (*a, b.clone()))
+        .collect::<Vec<_>>();
```

Comment:
it seems odd that the earlier case (where links are unresolved) only follows one link,
and here we follow all of the available (resolved) ones.
Is there a specific use case you have in mind?

I'd expect us to first resolve all unresolved links, and then offer all
(resolved) links, no matter whether we just resolved them, or whether they
were previously resolved.

I did not check whether there is some kind of guarantee by LSP that we can
rely on, but I doubt there is. So we probably want to handle the case where
we have both resolved and unresolved links exist.

(In theory, we could even cache the resolved links in `ctx.document_links`,
so we don't have to resolve them again. But I would not do that, since it
probably adds complexity for no significant gain).